File: core/src/manifest.ts
import { createHash } from "node:crypto";

export const MANIFEST_SCHEMA = "forensix/manifest/1" as const;
export const HASH_ALGORITHM = "sha-256" as const;

const MANIFEST_ENTRY_KEYS = [
  "path",
  "state",
  "unavailable_reason",
  "node_type",
  "file_kind",
  "selection_tier",
  "copied",
  "unclassified",
  "size",
  "mtime_ns",
  "hash_algorithm",
  "sha256",
  "link_target",
] as const;
const MANIFEST_HEADER_KEYS = [
  "manifest_schema",
  "source_kind",
  "selection_policy",
  "selection_policy_diff",
  "tier_2_included",
  "hash_algorithm",
  "evidence_set_digest",
  "working_copy_digest",
  "entry_count",
  "copied_entry_count",
  "profile_count",
  "unavailable_count",
  "unclassified_count",
] as const;

const MANIFEST_STATES = ["value", "absent", "unavailable"] as const;
const UNAVAILABLE_REASONS = [
  "permission_denied",
  "io_error",
  "changed_during_ingest",
  "parent_unavailable",
] as const;
const NODE_TYPES = ["file", "symlink", "socket", "dir", "absent"] as const;
const FILE_KINDS = [
  "database",
  "sidecar",
  "json",
  "image",
  "metadata",
  "liveness_evidence",
  "bulk_data",
  "ballast",
  "directory",
  "unclassified",
] as const;
const SELECTION_TIERS = ["tier_1", "tier_2", "tier_3", "unclassified"] as const;

export type ManifestState = (typeof MANIFEST_STATES)[number];
export type UnavailableReason = (typeof UNAVAILABLE_REASONS)[number];
export type NodeType = (typeof NODE_TYPES)[number];
export type FileKind = (typeof FILE_KINDS)[number];
export type SelectionTier = (typeof SELECTION_TIERS)[number];

export interface ManifestEntry {
  readonly path: string;
  readonly state: ManifestState;
  readonly unavailable_reason: UnavailableReason | null;
  readonly node_type: NodeType;
  readonly file_kind: FileKind;
  readonly selection_tier: SelectionTier;
  readonly copied: boolean;
  readonly unclassified: boolean;
  readonly size: number | null;
  readonly mtime_ns: string | null;
  readonly hash_algorithm: typeof HASH_ALGORITHM;
  readonly sha256: string | null;
  readonly link_target: string | null;
}

export interface ManifestHeader {
  readonly manifest_schema: typeof MANIFEST_SCHEMA;
  readonly source_kind: "USER_DATA_DIR";
  readonly selection_policy: "chrome-userdata/1";
  readonly selection_policy_diff: readonly [];
  readonly tier_2_included: boolean;
  readonly hash_algorithm: typeof HASH_ALGORITHM;
  readonly evidence_set_digest: string;
  readonly working_copy_digest: string;
  readonly entry_count: number;
  readonly copied_entry_count: number;
  readonly profile_count: number;
  readonly unavailable_count: number;
  readonly unclassified_count: number;
}

function record(value: unknown, name: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new TypeError(`${name} must be an object.`);
  }
  return value as Record<string, unknown>;
}

function assertExactKeys(
  value: Record<string, unknown>,
  expected: readonly string[],
  name: string,
): void {
  const actual = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  if (
    actual.length !== sortedExpected.length ||
    actual.some((key, index) => key !== sortedExpected[index])
  ) {
    throw new TypeError(`${name} has missing or unknown fields.`);
  }
}

function stringValue(value: unknown, name: string): string {
  if (typeof value !== "string") {
    throw new TypeError(`${name} must be a string.`);
  }
  return value;
}

function nullableString(value: unknown, name: string): string | null {
  return value === null ? null : stringValue(value, name);
}

function booleanValue(value: unknown, name: string): boolean {
  if (typeof value !== "boolean") {
    throw new TypeError(`${name} must be a boolean.`);
  }
  return value;
}

function nonnegativeInteger(value: unknown, name: string): number {
  if (!Number.isSafeInteger(value) || (value as number) < 0) {
    throw new TypeError(`${name} must be a nonnegative safe integer.`);
  }
  return value as number;
}

function nullableNonnegativeInteger(
  value: unknown,
  name: string,
): number | null {
  return value === null ? null : nonnegativeInteger(value, name);
}

function enumValue<const Values extends readonly string[]>(
  value: unknown,
  values: Values,
  name: string,
): Values[number] {
  if (typeof value !== "string" || !values.includes(value)) {
    throw new TypeError(`${name} has an unsupported value.`);
  }
  return value as Values[number];
}

function exactValue<const Value extends string>(
  value: unknown,
  expected: Value,
  name: string,
): Value {
  if (value !== expected) {
    throw new TypeError(`${name} must be ${expected}.`);
  }
  return expected;
}

export function sha256(data: NodeJS.ArrayBufferView | string): string {
  return createHash("sha256").update(data).digest("hex");
}

export function representationDigest(
  nodeType: Exclude<NodeType, "file">,
  linkTarget: string | null = null,
): string {
  if (nodeType === "symlink") {
    if (linkTarget === null) {
      throw new TypeError("A symlink representation needs target text.");
    }
    return sha256(Buffer.from(linkTarget, "utf8"));
  }

  return sha256(Buffer.from(nodeType, "utf8"));
}

function assertManifestInvariants(entry: ManifestEntry): void {
  if (
    entry.path.length === 0 ||
    entry.path.startsWith("/") ||
    entry.path.includes("\\") ||
    entry.path
      .split("/")
      .some((part) => part === "" || part === "." || part === "..")
  ) {
    throw new TypeError(
      `Manifest path is not canonical: ${JSON.stringify(entry.path)}`,
    );
  }
  if (entry.mtime_ns !== null && !/^-?[0-9]+$/.test(entry.mtime_ns)) {
    throw new TypeError(
      "A Manifest mtime must be a decimal nanosecond string.",
    );
  }
  if (entry.state === "unavailable") {
    if (
      entry.unavailable_reason === null ||
      entry.sha256 !== null ||
      entry.copied
    ) {
      throw new TypeError(
        "An unavailable entry needs a reason and cannot have a hash or copy.",
      );
    }
  } else if (entry.unavailable_reason !== null || entry.sha256 === null) {
    throw new TypeError(
      "An available or absent entry needs a hash and no unavailable reason.",
    );
  }
  if ((entry.state === "absent") !== (entry.node_type === "absent")) {
    throw new TypeError("Only an absent entry can have absent Node Type.");
  }
  if (entry.sha256 !== null && !/^[0-9a-f]{64}$/.test(entry.sha256)) {
    throw new TypeError(
      "A Manifest SHA-256 value must contain 64 lowercase hexadecimal digits.",
    );
  }
  if (entry.state === "absent") {
    if (
      entry.size !== null ||
      entry.mtime_ns !== null ||
      entry.sha256 !== representationDigest("absent")
    ) {
      throw new TypeError("An absent entry has an invalid representation.");
    }
  } else if (entry.state === "value" && entry.mtime_ns === null) {
    throw new TypeError("An available Source node needs an mtime.");
  }
  if (
    entry.node_type === "file" &&
    entry.state === "value" &&
    entry.size === null
  ) {
    throw new TypeError("An available file needs a size.");
  }
  if (
    (entry.node_type === "dir" || entry.node_type === "socket") &&
    entry.state === "value" &&
    (entry.size !== null ||
      entry.sha256 !== representationDigest(entry.node_type))
  ) {
    throw new TypeError("A directory or socket has an invalid representation.");
  }
  if (
    entry.node_type === "symlink" &&
    entry.state === "value" &&
    (entry.link_target === null ||
      entry.size !== Buffer.byteLength(entry.link_target, "utf8") ||
      entry.sha256 !== representationDigest("symlink", entry.link_target))
  ) {
    throw new TypeError("An available symlink has an invalid representation.");
  }
  if (entry.node_type !== "symlink" && entry.link_target !== null) {
    throw new TypeError("Only a symlink can contain target text.");
  }
  if (
    entry.copied &&
    (entry.state !== "value" ||
      entry.node_type !== "file" ||
      (entry.selection_tier !== "tier_1" && entry.selection_tier !== "tier_2"))
  ) {
    throw new TypeError(
      "Only an available Tier 1 or Tier 2 file can be copied.",
    );
  }
  if (
    entry.unclassified !==
    (entry.selection_tier === "unclassified" &&
      (entry.file_kind === "unclassified" || entry.file_kind === "directory"))
  ) {
    throw new TypeError("The unclassified fields disagree.");
  }
}

export function decodeManifestEntry(value: unknown): ManifestEntry {
  const object = record(value, "Manifest entry");
  assertExactKeys(object, MANIFEST_ENTRY_KEYS, "Manifest entry");
  const reason =
    object.unavailable_reason === null
      ? null
      : enumValue(
          object.unavailable_reason,
          UNAVAILABLE_REASONS,
          "unavailable_reason",
        );
  const entry: ManifestEntry = {
    path: stringValue(object.path, "path"),
    state: enumValue(object.state, MANIFEST_STATES, "state"),
    unavailable_reason: reason,
    node_type: enumValue(object.node_type, NODE_TYPES, "node_type"),
    file_kind: enumValue(object.file_kind, FILE_KINDS, "file_kind"),
    selection_tier: enumValue(
      object.selection_tier,
      SELECTION_TIERS,
      "selection_tier",
    ),
    copied: booleanValue(object.copied, "copied"),
    unclassified: booleanValue(object.unclassified, "unclassified"),
    size: nullableNonnegativeInteger(object.size, "size"),
    mtime_ns: nullableString(object.mtime_ns, "mtime_ns"),
    hash_algorithm: exactValue(
      object.hash_algorithm,
      HASH_ALGORITHM,
      "hash_algorithm",
    ),
    sha256: nullableString(object.sha256, "sha256"),
    link_target: nullableString(object.link_target, "link_target"),
  };
  assertManifestInvariants(entry);
  return entry;
}

export function assertManifestEntry(entry: ManifestEntry): void {
  decodeManifestEntry(entry);
}

export function decodeManifestHeader(value: unknown): ManifestHeader {
  const object = record(value, "Manifest header");
  assertExactKeys(object, MANIFEST_HEADER_KEYS, "Manifest header");
  if (
    !Array.isArray(object.selection_policy_diff) ||
    object.selection_policy_diff.length !== 0
  ) {
    throw new TypeError("selection_policy_diff must be an empty array.");
  }
  const header: ManifestHeader = {
    manifest_schema: exactValue(
      object.manifest_schema,
      MANIFEST_SCHEMA,
      "manifest_schema",
    ),
    source_kind: exactValue(object.source_kind, "USER_DATA_DIR", "source_kind"),
    selection_policy: exactValue(
      object.selection_policy,
      "chrome-userdata/1",
      "selection_policy",
    ),
    selection_policy_diff: [],
    tier_2_included: booleanValue(object.tier_2_included, "tier_2_included"),
    hash_algorithm: exactValue(
      object.hash_algorithm,
      HASH_ALGORITHM,
      "hash_algorithm",
    ),
    evidence_set_digest: stringValue(
      object.evidence_set_digest,
      "evidence_set_digest",
    ),
    working_copy_digest: stringValue(
      object.working_copy_digest,
      "working_copy_digest",
    ),
    entry_count: nonnegativeInteger(object.entry_count, "entry_count"),
    copied_entry_count: nonnegativeInteger(
      object.copied_entry_count,
      "copied_entry_count",
    ),
    profile_count: nonnegativeInteger(object.profile_count, "profile_count"),
    unavailable_count: nonnegativeInteger(
      object.unavailable_count,
      "unavailable_count",
    ),
    unclassified_count: nonnegativeInteger(
      object.unclassified_count,
      "unclassified_count",
    ),
  };
  if (
    !/^[0-9a-f]{64}$/.test(header.evidence_set_digest) ||
    !/^[0-9a-f]{64}$/.test(header.working_copy_digest)
  ) {
    throw new TypeError("Manifest digests must be SHA-256 values.");
  }
  if (header.copied_entry_count > header.entry_count) {
    throw new TypeError("copied_entry_count cannot exceed entry_count.");
  }
  return header;
}

export function canonicalManifestLine(entry: ManifestEntry): string {
  assertManifestEntry(entry);
  return JSON.stringify({
    path: entry.path,
    state: entry.state,
    unavailable_reason: entry.unavailable_reason,
    node_type: entry.node_type,
    file_kind: entry.file_kind,
    selection_tier: entry.selection_tier,
    copied: entry.copied,
    unclassified: entry.unclassified,
    size: entry.size,
    mtime_ns: entry.mtime_ns,
    hash_algorithm: entry.hash_algorithm,
    sha256: entry.sha256,
    link_target: entry.link_target,
  });
}

export function sortManifestEntries(
  entries: readonly ManifestEntry[],
): ManifestEntry[] {
  const sorted = [...entries].sort((left, right) =>
    Buffer.compare(
      Buffer.from(canonicalManifestLine(left), "utf8"),
      Buffer.from(canonicalManifestLine(right), "utf8"),
    ),
  );
  const paths = new Set<string>();
  for (const entry of sorted) {
    if (paths.has(entry.path)) {
      throw new TypeError(
        `Manifest path is duplicated: ${JSON.stringify(entry.path)}`,
      );
    }
    paths.add(entry.path);
  }
  return sorted;
}

export function manifestBytes(entries: readonly ManifestEntry[]): Buffer {
  if (entries.length === 0) {
    return Buffer.alloc(0);
  }

  const lines = sortManifestEntries(entries).map(canonicalManifestLine);
  return Buffer.from(`${lines.join("\n")}\n`, "utf8");
}

export function evidenceSetDigest(entries: readonly ManifestEntry[]): string {
  return sha256(manifestBytes(entries));
}

export function workingCopyDigest(entries: readonly ManifestEntry[]): string {
  return sha256(manifestBytes(entries.filter((entry) => entry.copied)));
}

File: collector/internal/conformance/manifest.go
package conformance

import (
	"bufio"
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path"
	"regexp"
	"sort"
	"strings"
)

var (
	decimalPattern = regexp.MustCompile(`^-?[0-9]+$`)
	digestPattern  = regexp.MustCompile(`^[0-9a-f]{64}$`)
)

// CanonicalLines encodes entries, validates unique paths, and byte-sorts the
// complete JSONL lines. This matches the Analyzer contract and LC_ALL=C sort.
func CanonicalLines(entries []ManifestEntry) ([]byte, error) {
	lines := make([][]byte, 0, len(entries))
	paths := make(map[string]bool, len(entries))
	for _, entry := range entries {
		if paths[entry.Path] {
			return nil, fmt.Errorf("duplicate Manifest path %q", entry.Path)
		}
		paths[entry.Path] = true
		if err := ValidateEntry(entry); err != nil {
			return nil, fmt.Errorf("Manifest entry %q: %w", entry.Path, err)
		}
		var line bytes.Buffer
		encoder := json.NewEncoder(&line)
		encoder.SetEscapeHTML(false)
		if err := encoder.Encode(entry); err != nil {
			return nil, fmt.Errorf("encode Manifest entry: %w", err)
		}
		// encoding/json always escapes these two valid Unicode characters;
		// JSON.stringify does not, and the contract permits only required JSON
		// escapes. Replace them to keep both implementations byte-identical.
		encoded := bytes.ReplaceAll(line.Bytes(), []byte(`\u2028`), []byte("\u2028"))
		encoded = bytes.ReplaceAll(encoded, []byte(`\u2029`), []byte("\u2029"))
		lines = append(lines, append([]byte(nil), encoded...))
	}
	sort.Slice(lines, func(i, j int) bool { return bytes.Compare(lines[i], lines[j]) < 0 })
	return bytes.Join(lines, nil), nil
}

func ValidateEntry(entry ManifestEntry) error {
	if entry.Path == "" || strings.HasPrefix(entry.Path, "/") || strings.Contains(entry.Path, `\`) || path.Clean(entry.Path) != entry.Path {
		return fmt.Errorf("path is not canonical")
	}
	for _, part := range strings.Split(entry.Path, "/") {
		if part == "" || part == "." || part == ".." {
			return fmt.Errorf("path is not canonical")
		}
	}
	if entry.HashAlgorithm != HashAlgorithm {
		return fmt.Errorf("unsupported hash algorithm %q", entry.HashAlgorithm)
	}
	if entry.Size != nil && *entry.Size < 0 {
		return fmt.Errorf("size is negative")
	}
	if entry.MTimeNS != nil && !decimalPattern.MatchString(*entry.MTimeNS) {
		return fmt.Errorf("mtime_ns is not a decimal nanosecond string")
	}
	if entry.SHA256 != nil && !digestPattern.MatchString(*entry.SHA256) {
		return fmt.Errorf("sha256 is not lowercase SHA-256")
	}
	if !validState(entry.State) || !validNodeType(entry.NodeType) || !validFileKind(entry.FileKind) || !validTier(entry.SelectionTier) {
		return fmt.Errorf("unsupported contract value")
	}

	if entry.State == StateUnavailable {
		if entry.UnavailableReason == nil || !validUnavailableReason(*entry.UnavailableReason) || entry.SHA256 != nil || entry.Copied {
			return fmt.Errorf("unavailable entry has an invalid reason, hash, or copy state")
		}
	} else if entry.UnavailableReason != nil || entry.SHA256 == nil {
		return fmt.Errorf("available or absent entry needs a hash and no unavailable reason")
	}
	if (entry.State == StateAbsent) != (entry.NodeType == NodeAbsent) {
		return fmt.Errorf("only an absent entry can have absent Node Type")
	}
	if entry.State == StateAbsent {
		if entry.Size != nil || entry.MTimeNS != nil || stringValue(entry.SHA256) != representationDigest(NodeAbsent, nil) {
			return fmt.Errorf("absent entry has an invalid representation")
		}
	} else if entry.State == StateValue && entry.MTimeNS == nil {
		return fmt.Errorf("available node needs an mtime")
	}
	if entry.State == StateValue && entry.NodeType == NodeFile && entry.Size == nil {
		return fmt.Errorf("available file needs a size")
	}
	if entry.State == StateValue && (entry.NodeType == NodeDir || entry.NodeType == NodeSocket) {
		if entry.Size != nil || stringValue(entry.SHA256) != representationDigest(entry.NodeType, nil) {
			return fmt.Errorf("directory or socket has an invalid representation")
		}
	}
	if entry.State == StateValue && entry.NodeType == NodeSymlink {
		if entry.LinkTarget == nil || entry.Size == nil || *entry.Size != int64(len([]byte(*entry.LinkTarget))) || stringValue(entry.SHA256) != representationDigest(NodeSymlink, entry.LinkTarget) {
			return fmt.Errorf("symlink has an invalid representation")
		}
	} else if entry.LinkTarget != nil {
		return fmt.Errorf("only a symlink can contain target text")
	}
	if entry.Copied && (entry.State != StateValue || entry.NodeType != NodeFile || (entry.SelectionTier != Tier1 && entry.SelectionTier != Tier2)) {
		return fmt.Errorf("only an available Tier 1 or Tier 2 file can be copied")
	}
	wantUnclassified := entry.SelectionTier == Unclassified && (entry.FileKind == KindUnclassified || entry.FileKind == KindDirectory)
	if entry.Unclassified != wantUnclassified {
		return fmt.Errorf("unclassified fields disagree")
	}
	return nil
}

func Digest(lines []byte) string {
	sum := sha256.Sum256(lines)
	return hex.EncodeToString(sum[:])
}

func WorkingCopyLines(entries []ManifestEntry) ([]byte, error) {
	copied := make([]ManifestEntry, 0, len(entries))
	for _, entry := range entries {
		if entry.Copied {
			copied = append(copied, entry)
		}
	}
	return CanonicalLines(copied)
}

func WriteJSON(path string, value any) error {
	file, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o600)
	if err != nil {
		return err
	}
	encoder := json.NewEncoder(file)
	encoder.SetEscapeHTML(false)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(value); err != nil {
		_ = file.Close()
		return err
	}
	return file.Close()
}

func WriteFile(path string, content []byte) error { return os.WriteFile(path, content, 0o600) }

func VerifySortedDigest(reader io.Reader) (string, error) {
	scanner := bufio.NewScanner(reader)
	var lines [][]byte
	for scanner.Scan() {
		lines = append(lines, append([]byte(nil), scanner.Bytes()...))
	}
	if err := scanner.Err(); err != nil {
		return "", err
	}
	sort.Slice(lines, func(i, j int) bool { return bytes.Compare(lines[i], lines[j]) < 0 })
	var canonical bytes.Buffer
	for _, line := range lines {
		canonical.Write(line)
		canonical.WriteByte('\n')
	}
	return Digest(canonical.Bytes()), nil
}

func representationDigest(node NodeType, linkTarget *string) string {
	value := string(node)
	if node == NodeSymlink && linkTarget != nil {
		value = *linkTarget
	}
	return Digest([]byte(value))
}

func stringValue(value *string) string {
	if value == nil {
		return ""
	}
	return *value
}

func validState(value ManifestState) bool {
	return value == StateValue || value == StateAbsent || value == StateUnavailable
}

func validUnavailableReason(value UnavailableReason) bool {
	return value == PermissionDenied || value == IOError || value == ChangedDuringIngest || value == ParentUnavailable
}

func validNodeType(value NodeType) bool {
	return value == NodeFile || value == NodeSymlink || value == NodeSocket || value == NodeDir || value == NodeAbsent
}

func validFileKind(value FileKind) bool {
	switch value {
	case KindDatabase, KindSidecar, KindJSON, KindImage, KindMetadata, KindLivenessEvidence, KindBulkData, KindBallast, KindDirectory, KindUnclassified:
		return true
	default:
		return false
	}
}

func validTier(value SelectionTier) bool {
	return value == Tier1 || value == Tier2 || value == Tier3 || value == Unclassified
}
